Cache Abbreviation entries after materializing them to speedup MCCAS replay

rastogishubham opened the pull request:
When looking at a trace of MCCAS replay, we noticed that a large chunk of time is spent in the function `visitDIEAttrs`

The reason for that is because, we materialize abbreviations every time we see a new DIE, which leads to decoding the same ULEB128 values that represent the form and attribute of a DIE for every DIE in the debug info section. We can materialize them once and cache them, which would reduce the time spent and give us some speedup.

We see some speedup as well:

MCCAS With Abbrev Caching
ninja clang  150.46s user 48.35s system 536% cpu 37.071 total

ninja clang  150.43s user 47.33s system 561% cpu 35.240 total

ninja clang  150.78s user 47.23s system 565% cpu 34.987 total

MCCAS Without Abbrev Caching
ninja clang  193.36s user 48.17s system 557% cpu 43.320 total

ninja clang  190.28s user 46.98s system 595% cpu 39.867 total

ninja clang  191.55s user 46.95s system 602% cpu 39.602 total

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3171,25 +3166,152 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
     return MaybeHasChildren.takeError();
 
-  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+  AbbrevDIETagAndChildren TagAndChildren;
```
It's better to avoid this "declare first, initialize later"  pattern:

```
  AbbrevDIETagAndChildren TagAndChildren = {*MaybeTag, *MaybeHasChildren};
```

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3090,9 +3090,26 @@ struct DIEVisitor {
+    dwarf::Form Form;
+    bool FormInDistinctData;
+    bool FormSizeRequiresDIEContents;
+    uint8_t FormSize;
```
`FormSizeRequiresDIEContents` and `FormSize` are mutually exclusive. Let's combine the two into an `optional<uint8_t> FormSize`.

rastogishubham replied:
Ahh yes, good idea!

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3166,153 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+    auto FormSize = FP.getDwarfOffsetByteSize();
+    return std::make_pair(false, FormSize);
+  }
+  case dwarf::DW_FORM_addrx3:
```
isn't addrx3's size == 3?

rastogishubham replied:
Yep my bad!

adrian-prantl replied:
Did you forget to update the patch?

rastogishubham replied:
Yeah minor mistake on my part, I made the change, but then I changed the function to no longer return a std::pair and rebasing made me lose the addrx3 move, I have fixed that now

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3166,153 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+      AbbrevDIECache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)].first;
+  StartTagCallback(TagAndChildren.Tag, AbbrevIdx);
+
+  if (auto E = visitDIEAttrs(DataReader, DIEData, AbbrevIdx))
```
Let's avoid querying the cache twice, we can replace `AbbrevIdx` here with the cached entry (its `.second` member)

felipepiovezan reviewed (commented):
exciting!

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3166,153 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  case dwarf::DW_FORM_ref_addr: {
+    auto FormSize =
+        (Form == dwarf::DW_FORM_addr) ? FP.AddrSize : FP.getRefAddrByteSize();
+    return std::make_pair(false, FormSize);
```
you can probably just write 
```suggestion
    return {false, FormSize};
```

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+}
+
+Error DIEVisitor::materializeAbbrevDIE(unsigned AbbrevIdx) {
+  constexpr auto AddrSize = 8;
```
Where does this come from? Should it be hardcoded?

rastogishubham replied:
For all of our purposes the AddressSize is always 8 bytes on MachO, if you see `DIEVisitor::visitDIEAttrs`, the same is true there

rastogishubham replied:
In fact, if you see `getAndSetDebugAbbrevOffsetAndSkip`, we have a check to make sure that the address size is 8

```    
if (AddressSize != 8)
      return createStringError(
          inconvertibleErrorCode(),
          "Address size is not 8 bytes, unsupported architecture for MCCAS!");
```

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+Error DIEVisitor::materializeAbbrevDIE(unsigned AbbrevIdx) {
+  constexpr auto AddrSize = 8;
+  constexpr auto FormParams =
+      dwarf::FormParams{4 /*Version*/, AddrSize, dwarf::DwarfFormat::DWARF32};
```
Is that also the version for DWARF 5?

adrian-prantl replied:
I just realized, what happens if you use MCCAS with a dwarf 2 binary? Is there a graceful degradation path?

felipepiovezan replied:
I guess we never really addressed the dwarf version issue.
We need to do that asap but separately

rastogishubham replied:
I agree with Felipe, the DWARF version is a separate issue and I will put up a separate patch for that

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  }
+
+  auto AbbrevDIECacheVal =
+      AbbrevDIECache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)];
```
we usually use the find() method because operator[] default-constructs the value and thus silently fails if there is no cache hit.

felipepiovezan replied:
This should be more evident once we disable the default constructor for that struct

rastogishubham replied:
I agree with Felipe, that is a seperate issue and I will put up a separate patch for that :)

rastogishubham replied:
I am confused, `AbbrevDIECache` is a `SmallVector`, not a `DenseMap`

adrian-prantl replied:
Ah nevermind.

adrian-prantl replied:
So how does this function work? It materializes all abbreviations (3299) every time an entry is needed? I must be missing something obvious...

rastogishubham replied:
The issue has been fixed, that was an oversight which did work when this patch was on top of the patch that switched materialization to iterative from recursive, but now I have a constructor in DIEVisitor to do the AbbrevEntryCache creation instead

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3172,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+static std::optional<uint8_t> getNonULEBFormSize(dwarf::Form Form,
+                                                 dwarf::FormParams FP) {
+  switch (Form) {
+  case dwarf::DW_FORM_addr:
+  case dwarf::DW_FORM_ref_addr: {
+    auto FormSize =
+        (Form == dwarf::DW_FORM_addr) ? FP.AddrSize : FP.getRefAddrByteSize();
+    return FormSize;
+  }
```
This line made me stop to figure out if there is something special going on here, but there is no reason to combine these `case`s?

```
  case dwarf::DW_FORM_addr:
    return FP.AddrSize
  case dwarf::DW_FORM_ref_addr:
    return FP.getRefAddrByteSize();
```

rastogishubham replied:
Yeah you are right, oversight on my part, sorry

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3172,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  AbbrevDIETagAndChildren TagAndChildren = {*MaybeTag, *MaybeHasChildren};
+  SmallVector<AbbrevDIEContents> AbbrevVector;
+  while (true) {
+    AbbrevDIEContents Contents;
```
Same idea as before, let's avoid uninitialized variables:

```
AbbrevDIEContents Contents = {*Attr, *Form, doesntDedup(*Form, *Attr}, 
                              getNonULEBFormSize(*Form, FormParams)}
```

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3087,12 +3087,29 @@ mccasformats::v1::loadDIETopLevel(DIETopLevelRef TopLevelRef) {
+  Error materializeAbbrevDIE(unsigned AbbrevIdx);
 
+  SmallVector<
+      std::pair<AbbrevDIETagAndChildren, SmallVector<AbbrevDIEContents>>>
```
Reading how this thing is used in subsequent commits, I think we can simplify the code a lot if we do away with the pair and fold it into `AbbrevDIETagAndChildren`

```
struct AbbrevEntry {
    dwarf::Tag Tag;
    bool HasChildren;
    SmallVector<AbbrevContent>  AbbrevContents;
};
```

(btw the "DIE" wording feels off and we could strip it?)

adrian-prantl replied:
Sounds plausible to me. @rastogishubham ?

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3172,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+    Contents.FormSize = FormSize;
+    AbbrevVector.push_back(Contents);
+  }
+  AbbrevDIECache.push_back(std::make_pair(TagAndChildren, AbbrevVector));
```
We should `std::move` this vector, since it is likely going to use memory allocations:

```
  AbbrevDIECache.emplace_back(TagAndChildren, std::move(AbbrevVector));
```

(This suggestion may vanish if you adopt my suggestion on the first commit)

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3172,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  auto AbbrevDIECacheVal =
+      AbbrevDIECache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)];
+  auto TagAndChildren = AbbrevDIECacheVal.first;
+  auto DIEContents = AbbrevDIECacheVal.second;
```
We're making a copy of the vector here.

felipepiovezan reviewed (approved):
Sorry, a bunch of things escaped me the first time I looked into this. Feel free to merge once you address these and Adrian's comments.
(I think the DWARF version issue should be handled separately, as it is a pre-existing problem)

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3087,12 +3087,50 @@ mccasformats::v1::loadDIETopLevel(DIETopLevelRef TopLevelRef) {
+    AbbrevEntryCache.reserve(AbbrevEntries.size());
+    for (unsigned I = 0; I < AbbrevEntries.size(); I++) {
+      if (Error E = materializeAbbrevDIE(encodeAbbrevIndex(I)))
+        report_fatal_error(std::move(E));
```
Note that we are crashing the compiler on invalid input.
I don't remember how this used to be done in previous iterations of this PR, but we could (maybe in a separate PR) move this out of DIEVisitor entirely; the DIEVisitor could take the already-cached abbreviations

rastogishubham replied:
Yes, once we move to the flat patch, we can do this materialization in `visitDIERef`

felipepiovezan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3155,25 +3186,127 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  auto &AbbrevEntryCacheVal =
+      AbbrevEntryCache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)];
+  // auto TagAndChildren = AbbrevDIECacheVal.first;
+  // ArrayRef<AbbrevContent> DIEContents = AbbrevDIECacheVal.second;
```
Please remove commented out code

felipepiovezan reviewed (approved):
LGTM

rastogishubham commented:
@swift-ci please test llvm

rastogishubham committed:
Introduce AbbrevContent and AbbrevEntry

rastogishubham committed:
Cache AbbrevEntries after materializing

This patch caches abbreviation entries after materializing them onces to
avoid rematerializing the same abbreviation entry for every DIE, thus
speeding up MCCAS replay
